File: DAA/knapsack01.java
import java.util.Scanner;

public class knapsack01 {

    public static int knapsack(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] dp = new int[n + 1][capacity + 1];

        // Build the dp table in a bottom-up manner
        for (int i = 0; i <= n; i++) {
            for (int w = 0; w <= capacity; w++) {
                if (i == 0 || w == 0) {
                    dp[i][w] = 0;
                } else if (weights[i - 1] <= w) {
                    dp[i][w] = Math.max(values[i - 1] + dp[i - 1][w - weights[i - 1]], dp[i - 1][w]);
                } else {
                    dp[i][w] = dp[i - 1][w];
                }
            }
        }

        // Store the result of the knapsack
        int result = dp[n][capacity];

        // Find the items that were included in the knapsack
        System.out.println("Selected items:");
        int w = capacity;
        for (int i = n; i > 0 && result > 0; i--) {
            if (result != dp[i - 1][w]) {
                System.out.println("Item " + i + " (Weight: " + weights[i - 1] + ", Value: " + values[i - 1] + ")");
                result -= values[i - 1];
                w -= weights[i - 1];
            }
        }

        // Return the maximum value we can get
        return dp[n][capacity];
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Ask user for the number of items
        System.out.print("Enter the number of items: ");
        int n = scanner.nextInt();

        // Initialize arrays for weights and values
        int[] weights = new int[n];
        int[] values = new int[n];

        // Input weights and values
        for (int i = 0; i < n; i++) {
            System.out.print("Enter weight and value for item " + (i + 1) + " (format: weight value): ");
            weights[i] = scanner.nextInt();
            values[i] = scanner.nextInt();
        }

        // Ask user for the capacity of the knapsack
        System.out.print("Enter the capacity of the knapsack: ");
        int capacity = scanner.nextInt();

        // Calculate and display the maximum value
        int maxValue = knapsack(weights, values, capacity);
        System.out.println("Maximum value: " + maxValue);

        scanner.close(); // Close the scanner to prevent resource leaks
    }
}

File: CPP Code/Knapsack01.cpp
#include <iostream>
#include <vector>
using namespace std;

// Function to solve 0-1 Knapsack problem using dynamic programming
int knapsackDP(int capacity, vector<int>& weights, vector<int>& values, int n) {
    // Create a 2D DP array with (n+1) rows and (capacity+1) columns
    vector<vector<int>> dp(n + 1, vector<int>(capacity + 1, 0));

    // Build the DP table
    for (int i = 1; i <= n; i++) {
        for (int w = 1; w <= capacity; w++) {
            if (weights[i - 1] <= w) {
                // Include the item or exclude it, choose the maximum value
                dp[i][w] = max(dp[i - 1][w], values[i - 1] + dp[i - 1][w - weights[i - 1]]);
            } else {
                // Exclude the item
                dp[i][w] = dp[i - 1][w];
            }
        }
    }

    // The maximum value is in the bottom-right cell
    return dp[n][capacity];
}

int main() {
    int n, capacity;
    cout << "Enter the number of items: ";
    cin >> n;
    cout << "Enter the capacity of the knapsack: ";
    cin >> capacity;

    vector<int> weights(n), values(n);
    cout << "Enter weights and values of each item:\n";
    for (int i = 0; i < n; i++) {
        cout << "Item " << i + 1 << " (Weight Value): ";
        cin >> weights[i] >> values[i];
    }

    int maxProfit = knapsackDP(capacity, weights, values, n);
    cout << "Maximum profit: " << maxProfit << endl;

    return 0;
}
